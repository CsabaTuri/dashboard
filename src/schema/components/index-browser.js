import React, { Component } from "react"
import { connect } from "react-redux"
import { query as q } from "faunadb"
import { Button, ButtonType } from "office-ui-fabric-react"

import { selectedIndex, selectedDatabase } from "../"
import { faunaClient } from "../../authentication"
import { watchForError } from "../../notifications"
import { monitorActivity, isBusy } from "../../activity-monitor"
import { InstancesList } from "../../dataset"
import { ReplEditor, evalQuery } from "../../repl"
import { KeyType } from "../../persistence/faunadb-wrapper"

export class IndexBrowser extends Component {

  constructor(props) {
    super(props)
    this.state = this.initialState(props)
    this.onChange = this.onChange.bind(this)
    this.onSubmit = this.onSubmit.bind(this)
    this.onSelectRef = this.onSelectRef.bind(this)
  }

  initialState(props) {
    return {
      terms: '""',
      queryFn: this.buildQuery(props)
    }
  }

  componentDidMount() {
    this.setState(this.initialState(this.props))
  }

  componentWillReceiveProps(next) {
    if (!this.props.index.equals(next.index)) {
      this.setState(this.initialState(next))
    }
  }

  buildQuery({ client, index, path }, terms = null) {
    if (!index.get("ref")) return

    if (index.get("terms").size !== 0) {
      if (!terms) return

      return options =>
        evalQuery(parsedTerms =>
          client.query(path, KeyType.SERVER, q.Paginate(
            q.Match(index.get("ref"), parsedTerms),
            options
          ))
        )(terms)
    }

    return (options) =>
      client.query(path, KeyType.SERVER, q.Paginate(
        q.Match(index.get("ref")),
        options
      ))
  }

  onChange(terms) {
    this.setState({ terms })
  }

  onSubmit() {
    this.setState({
      queryFn: this.buildQuery(
        this.props,
        this.state.terms
      )
    })
  }

  onSelectRef(ref) {
    const { client, path } = this.props

    return this.props.dispatch(
      monitorActivity(
        watchForError("Can't fetch selected instance", () =>
          client.query(path, KeyType.SERVER, q.Get(ref))
        )
      )
    )
  }

  render() {
    const { index, isBusy } = this.props
    const { terms, queryFn } = this.state

    return <div>
      {index.get("terms").size !== 0 ?
        <div className="ms-Grid-row">
          <div className="ms-Grid-col ms-u-sm12">
            <h3>Lookup</h3>
            <ReplEditor
              mode={ReplEditor.Mode.TEXT_FIELD}
              name="index-lookup-editor"
              value={terms}
              onChange={this.onChange}
              shortcuts={[{
                name: "search",
                bindKey: { win: "Enter", mac: "Enter" },
                exec: this.onSubmit
              }]} />

            <p className="ms-TextField-description">
              The contents of this field will be evaluated with the context of a repl and placed in
              the "terms" field of the index look up query.
            </p>

            <Button
              disabled={isBusy}
              buttonType={ButtonType.primary}
              onClick={this.onSubmit}>
              Search
            </Button>
          </div>
        </div> : null}

      <div className="ms-Grid-row">
        <div className="ms-Grid-col ms-u-sm12">
          <InstancesList query={queryFn} onSelectRef={this.onSelectRef} />
        </div>
      </div>
    </div>
  }
}

export default connect(
  state => ({
    client: faunaClient(state),
    index: selectedIndex(state),
    path: selectedDatabase(state).get("path"),
    isBusy: isBusy(state)
  })
)(IndexBrowser)
