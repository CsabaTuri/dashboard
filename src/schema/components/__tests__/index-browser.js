import React from "react"
import { Map, List } from "immutable"
import { shallow } from "enzyme"
import { shallowToJson } from "enzyme-to-json"
import { query as q } from "faunadb"

import { IndexBrowser } from "../index-browser"

jest.mock("../../../activity-monitor", () => ({
  monitorActivity: x => x
}))

jest.mock("../../../notifications", () => ({
  watchForError: () => Promise.resolve("mocked")
}))

describe("IndexBrowser Component", () => {
  let classIndex, index

  const render = (index) => shallow(
    <IndexBrowser dispatch={x => x} index={index} />
  )

  beforeEach(() => {
    classIndex = Map.of(
      "ref", q.Index("a-index"),
      "terms", List()
    )

    index = Map.of(
      "ref", q.Index("a-index"),
      "terms", List.of(
        Map.of("field", List.of("data", "name"))
      )
    )
  })

  it("browser a class index", () => {
    const comp = render(classIndex)
    expect(shallowToJson(comp)).toMatchSnapshot()
  })

  it("select a ref", () => {
    const comp = render(classIndex)
    return comp.find("Connect(Pagination)").props().onSelectRef(q.Ref("fake")).then(() => {
      comp.update()
      expect(shallowToJson(comp)).toMatchSnapshot()
    })
  })

  it("browser non class index", () => {
    const comp = render(index)
    expect(shallowToJson(comp)).toMatchSnapshot()

    comp.find("ReplEditor").simulate("change", "a name")
    comp.find("Button").simulate("click")
    expect(shallowToJson(comp)).toMatchSnapshot()
  })
})
