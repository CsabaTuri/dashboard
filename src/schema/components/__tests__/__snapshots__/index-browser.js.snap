// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IndexBrowser Component browser a class index 1`] = `
<div>
  <div
    className="ms-Grid-row"
  >
    <div
      className="ms-Grid-col ms-u-sm12"
    >
      <InstancesList
        onSelectRef={[Function]}
        query={[Function]}
      />
    </div>
  </div>
</div>
`;

exports[`IndexBrowser Component browser non class index 1`] = `
<div>
  <div
    className="ms-Grid-row"
  >
    <div
      className="ms-Grid-col ms-u-sm12"
    >
      <h3>
        Lookup
      </h3>
      <ReplEditor
        mode="editor form-field text-field"
        name="index-lookup-editor"
        onChange={[Function]}
        shortcuts={
          Array [
            Object {
              "bindKey": Object {
                "mac": "Enter",
                "win": "Enter",
              },
              "exec": [Function],
              "name": "search",
            },
          ]
        }
        value="\\"\\""
      />
      <p
        className="ms-TextField-description"
      >
        The contents of this field will be evaluated with the context of a repl and placed in the "terms" field of the index look up query.
      </p>
      <Button
        buttonType={1}
        onClick={[Function]}
      >
        Search
      </Button>
    </div>
  </div>
  <div
    className="ms-Grid-row"
  >
    <div
      className="ms-Grid-col ms-u-sm12"
    >
      <InstancesList
        onSelectRef={[Function]}
      />
    </div>
  </div>
</div>
`;

exports[`IndexBrowser Component browser non class index 2`] = `
<div>
  <div
    className="ms-Grid-row"
  >
    <div
      className="ms-Grid-col ms-u-sm12"
    >
      <h3>
        Lookup
      </h3>
      <ReplEditor
        mode="editor form-field text-field"
        name="index-lookup-editor"
        onChange={[Function]}
        shortcuts={
          Array [
            Object {
              "bindKey": Object {
                "mac": "Enter",
                "win": "Enter",
              },
              "exec": [Function],
              "name": "search",
            },
          ]
        }
        value="a name"
      />
      <p
        className="ms-TextField-description"
      >
        The contents of this field will be evaluated with the context of a repl and placed in the "terms" field of the index look up query.
      </p>
      <Button
        buttonType={1}
        onClick={[Function]}
      >
        Search
      </Button>
    </div>
  </div>
  <div
    className="ms-Grid-row"
  >
    <div
      className="ms-Grid-col ms-u-sm12"
    >
      <InstancesList
        onSelectRef={[Function]}
        query={[Function]}
      />
    </div>
  </div>
</div>
`;
